src/main/java/io/github/thebusybiscuit/slimefun4/core/services/MetricsService.java

```diff
@@ -0,0 +1,234 @@
+import java.util.logging.Level;
+
+import io.github.thebusybiscuit.slimefun4.implementation.SlimefunPlugin;
+import kong.unirest.HttpResponse;
+import kong.unirest.JsonNode;
+import kong.unirest.Unirest;
+import kong.unirest.UnirestException;
+import me.mrCookieSlime.Slimefun.api.Slimefun;
+import org.bukkit.plugin.Plugin;
+
+/**
+ * This Class represents a Metrics Service that sends data to https://bstats.org/
+ * This data is used to analyse the usage of this {@link Plugin}.
+ * <p>
+ * You can find more info in the README file of this Project on GitHub. <br>
+ * <b>Note:</b> To start the metrics you will need to be calling {@link #start()}
+ *
+ * @author TheBusyBiscuit
+ * @author WalshyDev
+ */
+public class MetricsService {
+
+    private static final String REPO_NAME = "MetricsModule";
+    private static final String GH_API = "https://api.github.com/repos/Slimefun/" + REPO_NAME;
+    private static final String GH_REPO_RELEASES = "https://github.com/Slimefun/" + REPO_NAME
+        + "/releases/download";
+
+    private final SlimefunPlugin plugin;
+    private final File metricFile;
+
+    private URLClassLoader moduleClassLoader;
+    private String metricVersion = null;
+
+    static {
+        Unirest.config()
+            .concurrency(2, 1)
+            .setDefaultHeader("User-Agent", "MetricsModule Auto-Updater")
+            .setDefaultHeader("Accept", "application/vnd.github.v3+json")
+            .enableCookieManagement(false)
+            .cookieSpec("ignoreCookies");
+    }
+
+    public MetricsService(SlimefunPlugin plugin) {
+        this.plugin = plugin;
+        this.metricFile = new File(plugin.getDataFolder(), REPO_NAME + ".jar");
+    }
+
+    /**
+     * This method loads the metric module and starts the metrics collection.
+     */
+    public void start() {
+        boolean newlyDownloaded = false;
+        if (!metricFile.exists()) {
+            info(REPO_NAME + " does not exist, downloading...");
+            if (!download(getLatestVersion())) {
+                plugin.getLogger().warning("Failed to start metrics as the file could not be downloaded.");
+                return;
+            }
+            newlyDownloaded = true;
+        }
+
+        try {
+            // Load the jar file into a child class loader using the SF PluginClassLoader
+            // as a parent.
+            moduleClassLoader = URLClassLoader.newInstance(new URL[] { metricFile.toURI().toURL() },
+                plugin.getClass().getClassLoader());
+            Class<?> cl = moduleClassLoader.loadClass("dev.walshy.sfmetrics.MetricsModule");
+
+            metricVersion = cl.getPackage().getImplementationVersion();
+
+            // If it has not been newly downloaded, auto-updates are on AND there's a new version
+            // then cleanup, download and start
+            if (!newlyDownloaded
+                && hasAutoUpdates()
+                && checkForUpdate(metricVersion)
+            ) {
```

Comment:
Also isnt this gonna check for updates twice if it finds one? Since it will re-run that method? Should probably set newlyDownloaded true inside download() to prevent that.

Reply:
Err, yeah could save a call.